ethcore/snapshot/benches/to_fat_rlps.rs

```diff
@@ -0,0 +1,93 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Benchmark snapshot::account::to_fat_rlps() which is a hot call during snapshots.
+
+use std::collections::HashSet;
+
+use account_db::AccountDB;
+use common_types::{
+	basic_account::BasicAccount,
+	snapshot::Progress
+};
+use criterion::{Criterion, criterion_group, criterion_main, black_box};
+use ethcore::test_helpers::new_temp_db;
+use ethereum_types::H256;
+use snapshot::test_helpers::to_fat_rlps;
+use tempdir::TempDir;
+use ethtrie::TrieDB;
+use trie_db::Trie;
+
+fn fat_rlps(c: &mut Criterion) {
+	let tempdir = TempDir::new("").unwrap();
+	let blockchain_db = new_temp_db(tempdir.path());
+
+	let mut state_rebuilder = snapshot::StateRebuilder::new(blockchain_db.key_value().clone(), journaldb::Algorithm::OverlayRecent);
+
+	// Chunk data collected from mainnet/ropsten around blocks 8.7M/6.8M (end of Oct '19). The data
+	// sizes represent roughly the 99-percentile of account sizes. It takes some effort to find
+	// accounts of representative size that are self-contained (i.e. do not have code in other
+	// chunks).
+	let chunks = vec![
+		// Ropsten
+		include_bytes!("./state-chunk-5279-0x2032dfb6ad93f1928dac70627a8767d2232568a1a7bf1c91ea416988000f8275.rlp").to_vec(),
+		// Ropsten
+		include_bytes!("./state-chunk-5905-0x104ff12a3fda9e0cb1aeef41fe7092982134eb116292c0eec725c32a815ef0ea.rlp").to_vec(),
+		// Ropsten
+		include_bytes!("./state-chunk-6341-0x3042ea62f982fd0cea02847ff0fd103a0beef3bb19389f5e77113c3ea355f803.rlp").to_vec(),
+		// Ropsten
+		include_bytes!("./state-chunk-6720-0x2075481dccdc2c4419112bfea2d09219a7223614656722a1a05a930baf2b0dd7.rlp").to_vec(),
+		// Mainnet
+		include_bytes!("./state-chunk-6933-0x104102770901b53230e78cfc8f6edce282eb21bfa00aa1c3543c79cb3402cf2d.rlp").to_vec(),
+	];
+
+	let flag = std::sync::atomic::AtomicBool::new(true);
+	for chunk in &chunks {
+		state_rebuilder.feed(&chunk, &flag).expect("feed fail");
+	}
+	let state_root = state_rebuilder.state_root();
+	let journal_db = state_rebuilder.finalize(123, H256::random()).expect("finalize fail");
+	let hashdb = journal_db.as_hash_db();
+	let account_trie = TrieDB::new(&hashdb, &state_root).expect("triedb has our root");
+	let account_iter = account_trie.iter().expect("there's a root in our trie");
+
+	for (idx, item) in account_iter.enumerate() {
+		let (account_key, account_data) = item.expect("data is the db is ok");
+		let account_hash = H256::from_slice(&account_key);
+		let basic_account: BasicAccount = rlp::decode(&*account_data).expect("rlp from disk is ok");
+		let account_db = AccountDB::from_hash(hashdb, account_hash);
+		let progress = Progress::new();
+		let mut used_code = HashSet::new();
```

Comment:
should `used_code` and `progress` be moved inside `b.iter`? not sure which case we're trying to measure

Reply:
How do you mean "which case"? You mean inside `to_fat_rlps()`? It's a big method and I'm not specifically targeting a particular branch but it started more from the observation of "wow, a loop in a loop in there, and I have no idea how many iterations this typically does" and then looking at the tests which are rather artificial I thought I'd collect some real data and see if call in its entirety is actually slow or not.
Turns out it's not, ~10us. I ran a few tests to check what typical loop counts are: outer loop once, sometimes twice, inner loop was like 5 times. Then I collected histograms of the # of bytes returned from here and it's mostly small and the 16 parts of the state seem mostly homogeneous.

Not sure if I have answered your question though! I don't think either needs to be in `b.iter()` but maybe I'm misunderstanding you.

Reply:
Yes, I mean inside `to_fat_rlps`. The perf is different depending on the content of `used_code` and you're passing the same reference to the `used_code` to `b.iter()` which runs the code in a loop.
I've got the following difference when moving the `used_code` inside `b.iter()`
```rust
to_fat_rlps, 5282 bytes, (0x1041…cf2d)                                                                               
                        time:   [6.2453 us 6.2689 us 6.2954 us]
                        change: [+18.545% +20.243% +23.000%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) high mild
  7 (7.00%) high severe

to_fat_rlps, 5908 bytes, (0x104f…f0ea)                                                                               
                        time:   [7.8668 us 7.8816 us 7.8975 us]
                        change: [+6.5079% +7.2548% +7.9737%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

to_fat_rlps, 6344 bytes, (0x2032…8275)                                                                               
                        time:   [7.8879 us 7.9030 us 7.9183 us]
                        change: [+10.102% +10.399% +10.715%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

to_fat_rlps, 6723 bytes, (0x2075…0dd7)                                                                               
                        time:   [9.8169 us 9.8328 us 9.8490 us]
                        change: [+8.6896% +9.0660% +9.4109%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

to_fat_rlps, 6936 bytes, (0x3042…f803)                                                                               
                        time:   [1.1720 us 1.1759 us 1.1808 us]
                        change: [+175.43% +177.79% +179.97%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe
 ```
the biggest difference is for the last chunk

Reply:
Interesting, good catch. Ok, let's move it inside then.

That last chunk is from mainnet btw, the others from ropsten.

Reply:
Yup, seeing the same effect:

```
     Running /Users/dvd/dev/parity/parity-ethereum/target/release/deps/to_fat_rlps-25811198c1c2ab93
to_fat_rlps, 5282 bytes, (0x1041…cf2d)
                        time:   [8.6684 us 8.8360 us 9.0185 us]
                        change: [+19.126% +22.714% +26.009%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

to_fat_rlps, 5908 bytes, (0x104f…f0ea)
                        time:   [10.474 us 10.588 us 10.710 us]
                        change: [+8.7095% +13.323% +17.690%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  9 (9.00%) high mild
  1 (1.00%) high severe

to_fat_rlps, 6344 bytes, (0x2032…8275)
                        time:   [10.401 us 10.536 us 10.687 us]
                        change: [+8.6806% +12.271% +15.996%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  4 (4.00%) high mild
  5 (5.00%) high severe

to_fat_rlps, 6723 bytes, (0x2075…0dd7)
                        time:   [12.859 us 13.089 us 13.350 us]
                        change: [+9.0429% +12.423% +16.180%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 14 outliers among 100 measurements (14.00%)
  7 (7.00%) high mild
  7 (7.00%) high severe

to_fat_rlps, 6936 bytes, (0x3042…f803)
                        time:   [2.0104 us 2.0549 us 2.1035 us]
                        change: [+113.79% +122.66% +131.16%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  6 (6.00%) high mild
  3 (3.00%) high severe
```

Reply:
Hmm, looking again I have doubts: in `chunk_state()` the `use_code` is instantiate once per part and re-used for all the millions of calls to `to_fat_rlps()` so maybe I got it right the first time? :/